src/shared/sleep-config.c

```diff
@@ -64,64 +86,26 @@ int parse_sleep_config(const char *verb, bool *ret_allow, char ***ret_modes, cha
                                         "Sleep\0", config_item_table_lookup, items,
                                         CONFIG_PARSE_WARN, NULL);
 
-        if (streq(verb, "suspend")) {
-                allow = allow_suspend != 0;
-
-                /* empty by default */
-                modes = TAKE_PTR(suspend_mode);
-
-                if (suspend_state)
-                        states = TAKE_PTR(suspend_state);
-                else
-                        states = strv_new("mem", "standby", "freeze");
-
-        } else if (streq(verb, "hibernate")) {
-                allow = allow_hibernate != 0;
-
-                if (hibernate_mode)
-                        modes = TAKE_PTR(hibernate_mode);
-                else
-                        modes = strv_new("platform", "shutdown");
-
-                if (hibernate_state)
-                        states = TAKE_PTR(hibernate_state);
-                else
-                        states = strv_new("disk");
-
-        } else if (streq(verb, "hybrid-sleep")) {
-                allow = allow_hybrid_sleep > 0 ||
-                        (allow_suspend != 0 && allow_hibernate != 0);
-
-                if (hybrid_mode)
-                        modes = TAKE_PTR(hybrid_mode);
-                else
-                        modes = strv_new("suspend", "platform", "shutdown");
-
-                if (hybrid_state)
-                        states = TAKE_PTR(hybrid_state);
-                else
-                        states = strv_new("disk");
-
-        } else if (streq(verb, "suspend-then-hibernate")) {
-                allow = allow_s2h > 0 ||
-                        (allow_suspend != 0 && allow_hibernate != 0);
-
-                modes = states = NULL;
-        } else
-                assert_not_reached("what verb");
-
-        if ((!modes && STR_IN_SET(verb, "hibernate", "hybrid-sleep")) ||
-            (!states && !streq(verb, "suspend-then-hibernate")))
+        /* use default values unless set */
+        if (!sc->suspend_states)
+                sc->suspend_states = strv_new("mem", "standby", "freeze");
+        if (!sc->hibernate_modes)
+                sc->hibernate_modes = strv_new("platform", "shutdown");
+        if (!sc->hibernate_states)
+                sc->hibernate_states = strv_new("disk");
+        if (!sc->hybrid_modes)
+                sc->hybrid_modes = strv_new("suspend", "platform", "shutdown");
+        if (!sc->hybrid_states)
+                sc->hybrid_states = strv_new("disk");
+        if (sc->hibernate_delay_sec == 0)
+                sc->hibernate_delay_sec = 180 * USEC_PER_MINUTE;
+
+        if (((streq(verb, "hibernate") && !sc->hibernate_modes) ||
+             (streq(verb, "hybrid-sleep") && !sc->hybrid_modes)) ||
+            (!streq(verb, "suspend-then-hibernate") && !sc->hibernate_states))
                 return log_oom();
```

Comment:
I'm not sure how this condition could be hit now but I tried to implement the same logic as the original code. If this should just be removed, please let me know.

Reply:
Before, the modes and states were set to something non-null only for certain verbs. But now you're not using the verb, and fields are set unconditionally, so this check should just make sure that everything that was supposed to be allocated is.

Reply:
@keszybz, I imagine there is a preferred approach for checks like this but would something like this work?

```C
        char *v;
        FOREACH_STRING(v, *sc->suspend_states,
                       *sc->hibernate_modes, *sc->hibernate_modes,
                       *sc->hybrid_modes, *sc->hybrid_states)
                if(!v)
                        return log_oom();
```